src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,468 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include <config.h>
+#include <array>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <assert.h>
+#include <libdrakvuf/json-util.h>
+#include <set>
+
+#include "ipt.h"
+#include "plugins/output_format.h"
+#include "private.h"
+
+
+/*
+ * Dump current IPT logs to disk and annotate them with custom PTWRITE packets
+ */
+static inline
+bool annotate_ipt(drakvuf_t drakvuf, drakvuf_trap_info_t* info, ipt* plugin, unsigned int vcpu_id, uint32_t* payloads, size_t num_payloads)
+{
+    uint64_t offset2;
+    uint64_t last_offset2;
+
+    uint64_t offset = info->regs->pt_offset;
+    uint64_t last_offset = plugin->vcpus[vcpu_id].last_offset;
+    bool ret = drakvuf_get_ipt_offset(drakvuf, vcpu_id, &offset2, &last_offset2);
+    int fail = 0;
+
+    uint8_t* buf = plugin->vcpus[vcpu_id].buf;
+    uint64_t size = plugin->vcpus[vcpu_id].size;
+    FILE* fd = plugin->vcpus[vcpu_id].fd;
+
+    plugin->vcpus[vcpu_id].last_offset = info->regs->pt_offset;
+
+    if (offset != offset2) {
+	printf("offset %llx %llx\n", (unsigned long long) offset, (unsigned long long) offset2);
+        throw -1;
+    }
+
+    /* if (!ret)
+    {
+        PRINT_DEBUG("annotate_ipt() failed to get ipt offset for vcpu %d\n", vcpu_id);
+        return false;
+    } */
+
+    PRINT_DEBUG("annotate_ipt() vCPU: %d IPT_CUR: %llx IPT_LAST: %llx\n", vcpu_id, (unsigned long long)offset, (unsigned long long)last_offset);
+
+    if (offset > last_offset)
+    {
+        fwrite(buf + last_offset, offset - last_offset, 1, fd);
+    }
+    else if (offset < last_offset)
+    {
+        fwrite(buf + last_offset, size - last_offset, 1, fd);
+        fwrite(buf, offset, 1, fd);
+    }
+    else
+    {
+        PRINT_DEBUG("annotate_ipt() called but no new IPT data is present\n");
+        fail = 1;
+    }
+
+    uint8_t ptwrite_packet[10] = {0x02, 0x32,};
+
+    for (size_t i = 0; i < num_payloads && !fail; i++)
+    {
+        uint32_t x1 = payloads[(i * 2) + 1];
+        uint32_t x2 = payloads[i * 2];
+
+        memcpy(&ptwrite_packet[2], &x1, sizeof(uint32_t));
+        memcpy(&ptwrite_packet[6], &x2, sizeof(uint32_t));
+        fwrite(ptwrite_packet, 10, 1, fd);
+    }
+
+    if (fail)
+    {
+        uint32_t x1 = PTW_ERROR_EMPTY;
+        uint32_t x2 = 0;
+
+        memcpy(&ptwrite_packet[2], &x1, sizeof(uint32_t));
+        memcpy(&ptwrite_packet[6], &x2, sizeof(uint32_t));
+        fwrite(ptwrite_packet, 10, 1, fd);
+    }
+
+    return true;
+}
+
+struct exec_fault_data
+{
+    ipt* plugin;
+    addr_t rip;
+};
+
+struct access_fault_result_t: public call_result_t
+{
+    access_fault_result_t() : call_result_t(), fault_va() {}
+
+    addr_t fault_va;
+};
+
+static event_response_t execute_faulted_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
```

Comment:
TODO this piece can be easily integrated into hyperbee (I think 90% of the code is already in place). So this plugin should be made much thinner.